File: experiments/verification-poc/import_trace.py
"""Untrusted external trace decoders. Every imported trace is replayed by verify.
Receipts detect accidental stale inputs; hashes do not authenticate a solver.
"""
import json
import re
from circuit import digest
from finite_model import need, value
from verify import FORMAT, verify


def sexprs(text):
    tokens=re.findall(r'\(|\)|[^\s()]+',text); pos=0
    def read(depth=0):
        nonlocal pos
        need(pos<len(tokens) and depth<128,'invalid S-expression')
        token=tokens[pos];pos+=1
        if token=='(':
            result=[]
            while pos<len(tokens) and tokens[pos]!=')': result.append(read(depth+1))
            need(pos<len(tokens),'unterminated S-expression');pos+=1;return result
        need(token!=')','unexpected closing parenthesis');return token
    result=[]
    while pos<len(tokens): result.append(read())
    return result


def smt_value(model,typ,raw):
    if typ=='Bool':
        need(raw in ('true','false'),'expected SMT Boolean');return raw=='true'
    width=model.width(typ)
    if type(raw) is str and re.fullmatch(r'#b[01]+',raw):
        need(len(raw)-2==width,'bit-vector width mismatch');number=int(raw[2:],2)
    elif type(raw) is str and re.fullmatch(r'#x[0-9a-fA-F]+',raw):
        need((len(raw)-2)*4==width,'bit-vector width mismatch');number=int(raw[2:],16)
    else:
        need(type(raw) is list and len(raw)==3 and raw[0]=='_' and re.fullmatch(r'bv[0-9]+',raw[1]) and raw[2]==str(width),'invalid SMT bit-vector')
        number=int(raw[1][2:]);need(number<2**width,'bit-vector out of range')
    if typ=='u8': return number
    need(number<len(model.enums[typ]),'unused enum encoding');return model.enums[typ][number]


def from_z3(model,text,bound):
    expressions=sexprs(text)
    need(len(expressions)==2 and expressions[0]=='sat' and type(expressions[1]) is list,'expected sat and one get-value result')
    values={}
    for pair in expressions[1]:
        need(type(pair) is list and len(pair)==2 and type(pair[0]) is str and pair[0] not in values,'invalid/duplicate SMT value');values[pair[0]]=pair[1]
    names={f's{i}_f_{f}' for i in range(bound+1) for f in model.fields}
    need(set(values)==names,'SMT values do not match BMC variables')
    return [{f:smt_value(model,t,values[f's{i}_f_{f}']) for f,t in model.fields.items()} for i in range(bound+1)]


def from_tlc(model,text):
    # TLC v1.8.0 _JsonTrace: a set of <<level, record-of-spec-variables>>.
    def unique(pairs):
        result={}
        for k,v in pairs:
            need(k not in result,'duplicate JSON key');result[k]=v
        return result
    raw=json.loads(text,object_pairs_hook=unique)
    need(type(raw) is dict and raw.get('vars')==['state'],'unexpected TLC variable set')
    graph=raw.get('counterexample');need(type(graph) is dict,'missing TLC counterexample')
    entries=graph.get('state');need(type(entries) is list and 0<len(entries)<=4096,'invalid TLC trace states')
    levels={}
    for entry in entries:
        need(type(entry) is list and len(entry)==2 and type(entry[0]) is int and entry[0]>0 and entry[0] not in levels,'invalid/duplicate TLC level')
        record=entry[1];need(type(record) is dict and set(record)=={'state'},'unexpected TLC state variables')
        fields=record['state'];need(type(fields) is dict and set(fields)=={'f_'+f for f in model.fields},'TLC fields mismatch')
        state={}
        for f,t in model.fields.items():
            v=fields['f_'+f]
            if t in model.enums:
                need(type(v) is str and v.startswith(t+'.'),'nominal enum mismatch');v=v[len(t)+1:]
            state[f]=v
        model.valid_state(state);levels[entry[0]]=state
    need(set(levels)==set(range(1,len(levels)+1)),'noncontiguous TLC trace')
    return [levels[i] for i in range(1,len(levels)+1)]


def receipt(model,backend,query,raw,bound=None):
    return {'format':'oak-trace-receipt-1','semantic_digest':model.digest,'backend':backend,'query_sha256':digest(query),'output_sha256':digest(raw),'bound':bound}


def import_trace(model,backend,query,raw,record,bound=None):
    need(record==receipt(model,backend,query,raw,bound),'stale/mismatched trace receipt')
    # Recompute the exact projection; a receipt alone is not a semantic check.
    if backend=='z3':
        from projections import bmc
        need(type(bound) is int and query==bmc(model,bound,True),'wrong BMC query')
        states=from_z3(model,raw,bound)
    elif backend=='tlc':
        import tempfile
        from pathlib import Path
        from projections import emit
        with tempfile.TemporaryDirectory() as folder:
            emit(model,folder)
            expected=(Path(folder)/'Model.tla').read_text()+'\n'+(Path(folder)/'Model.cfg').read_text()
        need(bound is None and query==expected,'wrong TLC specification/configuration')
        states=from_tlc(model,raw)
    else: raise ValueError('unsupported trace backend')
    # BMC can continue after its first violation. Replay the whole candidate
    # first, then retain the shortest prefix ending at an unsafe state.
    for state in states: model.valid_state(state)
    need(value(model.terms['initial'],{'s':states[0]}),'invalid initial state')
    for s,t in zip(states,states[1:]): need(s==t or value(model.terms['step'],{'s':s,'t':t}),'illegal transition')
    end=next((i for i,s in enumerate(states) if not value(model.terms['invariant'],{'s':s})),None)
    need(end is not None,'no safety violation')
    cert={'format':FORMAT,'semantic_digest':model.digest,'kind':'trace','states':states[:end+1]}
    verify(model,cert)
    return cert

File: experiments/verification-poc/import_trace.go
package main

import (
	"fmt"
	"reflect"
	"strconv"
	"strings"
)

type Receipt struct {
	Format    string `json:"format"`
	Digest    string `json:"semantic_digest"`
	Backend   string `json:"backend"`
	QuerySHA  string `json:"query_sha256"`
	OutputSHA string `json:"output_sha256"`
	Bound     *int   `json:"bound"`
}

func receipt(m *Model, backend, query, raw string, bound *int) Receipt {
	return Receipt{"oak-go-trace-receipt-1", m.Digest, backend, digest(query), digest(raw), bound}
}
func sexprs(text string) (out []any, err error) {
	if len(text) > 20_000_000 {
		return nil, fmt.Errorf("solver output limit")
	}
	tokens := strings.Fields(strings.NewReplacer("(", " ( ", ")", " ) ").Replace(text))
	pos := 0
	var read func(int) (any, error)
	read = func(depth int) (any, error) {
		if pos >= len(tokens) || depth > 128 {
			return nil, fmt.Errorf("invalid S-expression")
		}
		token := tokens[pos]
		pos++
		if token == ")" {
			return nil, fmt.Errorf("unexpected closing parenthesis")
		}
		if token != "(" {
			return token, nil
		}
		a := []any{}
		for pos < len(tokens) && tokens[pos] != ")" {
			v, e := read(depth + 1)
			if e != nil {
				return nil, e
			}
			a = append(a, v)
		}
		if pos >= len(tokens) {
			return nil, fmt.Errorf("unterminated S-expression")
		}
		pos++
		return a, nil
	}
	for pos < len(tokens) {
		v, e := read(0)
		if e != nil {
			return nil, e
		}
		out = append(out, v)
	}
	return out, nil
}
func smtValue(m *Model, t string, raw any) any {
	if t == "Bool" {
		s, ok := raw.(string)
		demand(ok && (s == "true" || s == "false"), "expected SMT Boolean")
		return s == "true"
	}
	width := m.width(t)
	var n int64
	var e error
	if s, ok := raw.(string); ok && strings.HasPrefix(s, "#b") {
		demand(len(s)-2 == width, "bit-vector width mismatch")
		n, e = strconv.ParseInt(s[2:], 2, 32)
	} else if ok && strings.HasPrefix(s, "#x") {
		demand((len(s)-2)*4 == width, "bit-vector width mismatch")
		n, e = strconv.ParseInt(s[2:], 16, 32)
	} else {
		a := array(raw)
		demand(len(a) == 3 && str(a[0]) == "_" && strings.HasPrefix(str(a[1]), "bv") && str(a[2]) == strconv.Itoa(width), "invalid bit-vector")
		n, e = strconv.ParseInt(str(a[1])[2:], 10, 32)
	}
	demand(e == nil && n >= 0 && n < int64(1<<width), "invalid bit-vector value")
	if t == "u8" {
		return int(n)
	}
	demand(int(n) < len(m.Enums[t]), "unused enum encoding")
	return m.Enums[t][n]
}
func fromZ3(m *Model, text string, bound int) (states []State, err error) {
	defer func() {
		if e := recover(); e != nil {
			states = nil
			err = fmt.Errorf("invalid Z3 trace: %v", e)
		}
	}()
	demand(bound >= 0 && bound <= 64, "invalid bound")
	a, e := sexprs(text)
	if e != nil {
		return nil, e
	}
	demand(len(a) == 2 && str(a[0]) == "sat", "expected SAT plus get-value result")
	values := map[string]any{}
	for _, v := range array(a[1]) {
		pair := array(v)
		demand(len(pair) == 2, "invalid value pair")
		key := str(pair[0])
		_, present := values[key]
		demand(!present, "duplicate SMT variable")
		values[key] = pair[1]
	}
	demand(len(values) == (bound+1)*len(m.Fields), "SMT variable set mismatch")
	for i := 0; i <= bound; i++ {
		s := State{}
		for _, f := range m.Fields {
			key := fmt.Sprintf("s%d_f_%s", i, f.Name)
			v, ok := values[key]
			demand(ok, "missing SMT variable")
			s[f.Name] = smtValue(m, f.Type, v)
		}
		states = append(states, s)
	}
	return states, nil
}
func fromTLC(m *Model, text string) (states []State, err error) {
	defer func() {
		if e := recover(); e != nil {
			states = nil
			err = fmt.Errorf("invalid TLC trace: %v", e)
		}
	}()
	var raw map[string]any
	if e := decode([]byte(text), &raw); e != nil {
		return nil, e
	}
	vars := array(raw["vars"])
	demand(len(vars) == 1 && str(vars[0]) == "state", "unexpected TLC variable set")
	graph, ok := raw["counterexample"].(map[string]any)
	demand(ok, "missing TLC counterexample")
	entries := array(graph["state"])
	demand(len(entries) > 0 && len(entries) <= 4096, "invalid TLC trace size")
	levels := map[int]State{}
	for _, v := range entries {
		entry := array(v)
		demand(len(entry) == 2, "invalid TLC state entry")
		level, e := number(entry[0])
		demand(e == nil && level > 0 && levels[level] == nil, "invalid/duplicate TLC level")
		record, ok := entry[1].(map[string]any)
		demand(ok && len(record) == 1, "unexpected TLC state variables")
		fields, ok := record["state"].(map[string]any)
		demand(ok && len(fields) == len(m.Fields), "TLC fields mismatch")
		s := State{}
		for _, f := range m.Fields {
			v, ok := fields["f_"+f.Name]
			demand(ok, "missing TLC field")
			if m.Enums[f.Type] != nil {
				txt := str(v)
				demand(strings.HasPrefix(txt, f.Type+"."), "nominal enum mismatch")
				v = strings.TrimPrefix(txt, f.Type+".")
			}
			s[f.Name] = v
		}
		demand(m.valid(s) == nil, "invalid typed TLC state")
		levels[level] = s
	}
	for i := 1; i <= len(levels); i++ {
		demand(levels[i] != nil, "noncontiguous TLC levels")
		states = append(states, levels[i])
	}
	return states, nil
}
func importTrace(m *Model, backend, query, raw string, r Receipt, bound *int) (*Certificate, error) {
	if !reflect.DeepEqual(r, receipt(m, backend, query, raw, bound)) {
		return nil, fmt.Errorf("stale or mismatched trace receipt")
	}
	var states []State
	var e error
	switch backend {
	case "z3":
		if bound == nil {
			return nil, fmt.Errorf("missing BMC bound")
		}
		expected, err := bmc(m, *bound, true)
		if err != nil || query != expected {
			return nil, fmt.Errorf("wrong BMC query")
		}
		states, e = fromZ3(m, raw, *bound)
	case "tlc":
		files := project(m)
		if bound != nil || query != files["Model.tla"]+"\n"+files["Model.cfg"] {
			return nil, fmt.Errorf("wrong TLC specification/configuration")
		}
		states, e = fromTLC(m, raw)
	default:
		return nil, fmt.Errorf("unsupported trace backend")
	}
	if e != nil {
		return nil, e
	}
	if e := replay(m, states); e != nil {
		return nil, e
	}
	for i, s := range states {
		if !truth(m.Terms["invariant"], s, nil) {
			c := &Certificate{Format: evidenceFormat, Digest: m.Digest, Kind: "trace", States: states[:i+1]}
			_, e := verify(m, c)
			return c, e
		}
	}
	return nil, fmt.Errorf("trace has no safety violation")
}
